src/drivers/mediatek/mt8186/afe-memif.c

```diff
@@ -0,0 +1,719 @@
+	memif->irq_id = irq_id;
+	memif->dma_base = dma_addr;
+	memif->dma_size = dma_size;
+	memif->direction = direction;
+	/* TODO risk, it may has sync problems with DAI comp */
+	memif->rptr = 0;
+	memif->wptr = 0;
+	memif->period_size = config->elem_array.elems[0].size;
+
+	/* get dai's config setting from afe driver */
+	ret = afe_dai_get_config(memif->afe, dai_id, &memif->channel, &memif->rate, &memif->format);
+	if (ret < 0)
+		return ret;
+	/* set the afe memif parameters */
+	ret = afe_memif_set_params(memif->afe, memif->memif_id, memif->channel, memif->rate,
+				   memif->format);
+	if (ret < 0)
+		return ret;
+	ret = afe_memif_set_addr(memif->afe, memif->memif_id, memif->dma_base, memif->dma_size);
+	if (ret < 0)
+		return ret;
+	channel->status = COMP_STATE_PREPARE;
+
+	return 0;
+}
+
+static int memif_probe(struct dma *dma)
+{
+	int channel;
+	int ret;
+	struct mtk_base_afe *afe = afe_get();
+	struct afe_memif_dma *memif;
+
+	if (!dma || dma->chan) {
+		tr_err(&memif_tr, "MEMIF: Repeated probe");
+		return -EEXIST;
+	}
+
+	/* do afe driver probe */
+	ret = afe_probe(afe);
+	if (ret < 0) {
+		tr_err(&memif_tr, "MEMIF: afe_probe fail:%d", ret);
+		return ret;
+	}
+
+	dma->chan = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+			    dma->plat_data.channels * sizeof(struct dma_chan_data));
+	if (!dma->chan) {
+		tr_err(&memif_tr, "MEMIF: Probe failure, unable to allocate channel descriptors");
+		return -ENOMEM;
+	}
+
+	for (channel = 0; channel < dma->plat_data.channels; channel++) {
+		dma->chan[channel].dma = dma;
+		/* TODO need divide to UL and DL for different index */
+		dma->chan[channel].index = channel;
+
+		memif = rzalloc(SOF_MEM_ZONE_SYS_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+				sizeof(struct afe_memif_dma));
+		if (!memif) {
+			tr_err(&memif_tr, "afe-memif: %d channel %d private data alloc failed",
+			       dma->plat_data.id, channel);
+			goto out;
+		}
+
+		memif->afe = afe;
+		memif->memif_id = channel;
+		dma_chan_set_data(&dma->chan[channel], memif);
+	}
+	return 0;
+
+out:
+	if (dma->chan) {
+		for (channel = 0; channel < dma->plat_data.channels; channel++)
+			rfree(dma_chan_get_data(&dma->chan[channel]));
+		rfree(dma->chan);
+		dma->chan = NULL;
+	}
+
+	afe_remove(afe);
+
+	return -ENOMEM;
+}
+
+static int memif_remove(struct dma *dma)
+{
+	int channel;
+	struct mtk_base_afe *afe = afe_get();
+
+	if (!dma->chan) {
+		tr_err(&memif_tr, "MEMIF: remove called without probe, it's a no-op");
+		return 0;
+	}
+	for (channel = 0; channel < dma->plat_data.channels; channel++) {
+		/* TODO Disable HW requests for this channel */
+
+		rfree(dma_chan_get_data(&dma->chan[channel]));
+	}
+	rfree(dma->chan);
+	dma->chan = NULL;
+
+	return afe_remove(afe);
+}
+
+static int memif_interrupt(struct dma_chan_data *channel, enum dma_irq_cmd cmd)
+{
+	int ret;
+	unsigned int period;
+	unsigned int sample_size;
+	struct afe_memif_dma *memif;
+	struct mtk_base_afe *afe = afe_get();
+
+	if (channel->status == COMP_STATE_INIT)
+		return 0;
```

Comment:
shouldnt this be an error? can you give a bit more detail here maybe in a comment about why we need this check?

Reply:
DMA should be used after dma_channel_get, after dma_channel_get, the channel->status will change to COMP_STATE_READY.